### ISSUE TITLE
[BUG] New cattle impersonation token for every call to downstream cluster

### ISSUE THREAD

**Issue by rmdvb** [rancher/rancher#43717]
**Rancher Server Setup**
- Rancher version: 2.7.9 (upgraded from 2.6.13 which was in the past upgraded from 2.6.8 and 2.6.3)
- Installation option (Docker install/Helm Chart):
   - If Helm Chart, Kubernetes Cluster and version (RKE1, RKE2, k3s, EKS, etc):
   - Helm chart: v2.7.9
   - Using EKS clusters imported as generic cluster
- Proxy/Cert Details:
   - Managed by cert-manager (private CA)

**Information about the Cluster**
- Kubernetes version: 1.25 (Rancher server on K8s v1.24)
- Cluster Type Downstream
   - Downstream, generic imported cluster.

**User Information**
- What is the role of the user logged in? Admin

**Describe the bug**
Rancher server on our main cluster seems to be running fine. Rancher-agent on the client seems to be running fine.
However for every call made through kubectl a new token secret is created for the service account

**To Reproduce**
- Upgrade from Rancher 2.6.13 to 2.7.9.

**Result**
For every Kubectl call made through Rancher a new token is created logs from Rancher server:
`[INFO] EnsureSecretForServiceAccount: waiting for secret [cattle-impersonation-u-******-token-9tq84] to be populated with token`

This repeates for every call. Sometimes calls finish other times they throw an error message about not being able to locate the secret for the service account needed.

`$ k get po --kubeconfig <rancher-kubeconfig>.yaml`
`No resources found in default namespace.`

`$ k get po --kubeconfig <rancher-kubeconfig>.yaml`
`No resources found in default namespace.`

`$ k get po --kubeconfig <rancher-kubeconfig>.yaml`
`Error from server (InternalError): an error on the server ("unable to create impersonator account: error setting up impersonation for user u-********: failed to get secret for service account: cattle-impersonation-system/cattle-impersonation-u-******, error: timed out waiting for the condition") has prevented the request from succeeding (get pods)`

If we look at the secrets on the downstream cluster we see a new secret was created for each call.

This results in our teams losing connection to the cluster and being unable to commit changes succesfully.

**Expected Result**
To re-use the same token after creation

**Additional context**
We have upgraded from v2.6.13 to v2.7.9, since then this error occured.
Seems like the message we see comes from here: https://github.com/rancher/rancher/commit/8f4d60c2228de4672a469d07213778ce53df9433#diff-8a853a257f4c9e8e5812b706a4f6a56eaa32ff67e57f9ac5ac4f78f1e1659f04R65

**Comment by Mart-Kuc** [comment]
We had similar problem. During upgrade rancher from `v2.7.6` to `v2.7.9`, in `cattle-impersonation-system` namespace, thousands of secrets (cattle-impersonation-user****) were generated until server crashed due to OOM killed.

In logs we also found `[INFO] EnsureSecretForServiceAccount: waiting for secret [cattle-impersonation-u-xxxxxx] to be populated with token`

**Additional information:**
- K8S version  v1.25.16

**Comment by rmdvb** [comment]
Did a downgrade solve this issue for you?

**Comment by Mart-Kuc** [comment]
> Did a downgrade solve this issue for you?
> 
> Seems like this might be a similar issue: #17456

I try to describe it more precisely what we did and how we 'fixed' it:
-  After upgrade to `v2.7.9` (which was succeed) secrets were generated in huge amounts until OOM killer occurred.
- We had to increase memory (RAM) on our masters, then masters were able to start
- And we had to manually deleted all secrets in `cattle-impersonation-system` namespace

After this action cluster was up and running and looks stable.

We tried upgrade 3 clusters and all of this had same problem

**Comment by rmdvb** [comment]
Did you delete the secrets on the downstream-cluster or on the Rancher server?

**Comment by Mart-Kuc** [comment]
Only on downstream-cluster

**Comment by rmdvb** [comment]
Thank you for your help Mart-Kuc!
Unfortunately this doesn't solve the issue for us.

**Comment by daleckystepan** [comment]
I hit the same  issue. I only did update of Kubernetes v1.25 to v1.26.11 using Rancher v2.7.9. No more changes.
Thousands of secrets have been created during and after the update.

**Comment by feliipecosta** [comment]
Same here, i've upgraded the upstream cluster from 2.7.4 to 2.7.9 and downstream clusters started to create impersonation secrets wich huge increased the ETCD size, i suspect that when upstream cluster is requesting a new impersonation token secret it is getting timed out and the request restart the secret creation, i found many of this on log:

E0108 21:10:08.585173       1 machineset_controller.go:890] "Unable to retrieve Node status" err="failed to create cluster accessor: error creating dynamic rest mapper for remote cluster \"ns-xpto\": an error on the server (\"unable to create impersonator account: error getting service account token: error getting secret: error ensuring secret for service account [cattle-impersonation-system:cattle-impersonation-u-brlgzerlri]: timed out waiting for the condition\")...

**Comment by feliipecosta** [comment]
As you can see on the output, there is ~380000 secrets on this downstream cluster ETCD.

```
> kubectl -n kube-system exec etcd-namespace-93d65ec6-64tjc -- sh -c "ETCDCTL_ENDPOINTS='https://127.0.0.1:2379/' ETCDCTL_CACERT='/var/lib/rancher/rke2/server/tls/etcd/server-ca.crt' ETCDCTL_CERT='/var/lib/rancher/rke2/server/tls/etcd/server-client.crt' ETCDCTL_KEY='/var/lib/rancher/rke2/server/tls/etcd/server-client.key' ETCDCTL_API=3 etcdctl get '' --prefix --keys-only | grep -v ^$" | awk -F '/'  '{ h[$3]++ } END {for (k in h) print h[k], k}' | sort -nr 

389666 secrets
798 events
364 replicasets
```

**Comment by daleckystepan** [comment]
I checked multiple clusters in different versions and each of them has duplicated secrets for one particular user. So it is probably long lasting bug. Update of cluster makes it significantly worse. tens thousands of secrets and then OOM of API server or/and etcd.

Strange thing is that
`kubectl delete secrets -n cattle-impersonation-system --all` helps and it settles down after that. Still increasing but slowly.

**Comment by renatogalera** [comment]
> I checked multiple clusters in different versions and each of them has duplicated secrets for one particular user. So it is probably long lasting bug. Update of cluster makes it significantly worse. tens thousands of secrets and then OOM of API server or/and etcd.
> 
> Strange thing is that `kubectl delete secrets -n cattle-impersonation-system --all` helps and it settles down after that. Still increasing but very slowly.
> 
> btw: We use AD for user auth. No idea if it is connected with this bug or not.


Hi, you can do this directly in etcd, this takes 2 seconds.

Note: It got better after upgrading to version 2.8.0, but I still see more secrets than before

```
kubectl exec -ti -n kube-system $(kubectl get pods -n kube-system -l component=etcd -o jsonpath="{.items[0].metadata.name}") -- bash
```

Inside pod, execute

```
etcdctl --cert /var/lib/rancher/rke2/server/tls/etcd/server-client.crt --key /var/lib/rancher/rke2/server/tls/etcd/server-client.key --endpoints https://127.0.0.1:2379/ --cacert /var/lib/rancher/rke2/server/tls/etcd/server-ca.crt del --prefix /registry/secrets/cattle-impersonation-system
```

**Comment by daleckystepan** [comment]
Anyone using ArgoCD with user-abcde | Default Admin account? Or any other app that uses admin local account heavily?

I am not sure if ti is connected or not. I am just trying to figure out how to decrease likelihood that it will happen again.

**Comment by renatogalera** [comment]
This worsened after our infrastructure became slow.

What I observed was an attempt to create a token, with a timeout being reached, and yet, that token was still created.

This led to a looping of token creation, which might be a bug.

**Comment by MKlimuszka** [comment]
Assigning to team/collie as this is impersonation related.

**Comment by ikogan** [comment]
I wonder if [this issue](https://github.com/rancher/rancher/issues/44347) is related. We just experienced this for our Portworx service account. Deleting all of the tokens for that account brought the cluster back to life and stopping the Portworx pods caused the tokens to stop being generated. This randomly started on 5/31 but we can't determine any clear cause.

**Comment by sosimon** [comment]
We're running into this also, at least the symptom of having thousands of secrets (service account tokens) for `cattle-impersonation-user*`, leading to a few OOM issues. We are on Rancher v2.8.1.

**Comment by moio** [comment]
@bigkevmcd to my non-expert eyes this is not expected behavior - or is it? Can you or somebody else from Collie confirm?

**Comment by moio** [comment]
(potentially @raulcabello)

**Comment by bigkevmcd** [comment]
@moio We're working on a fix for this, a  change has landed which will reduce it somewhat, but we need a more invasive change to completely prevent it happening.

**Comment by greeneggies** [comment]
closing in favor of #47126

**Comment by greeneggies** [comment]
https://github.com/rancher/rancher/issues/44347 was fixed in 2.10, 2.9, and 2.8 as a partial fix - the issue linked above is being worked on to wrap it up completely, which should encapsulate th problem outlined here

**Comment by samjustus** [comment]
https://github.com/rancher/rancher/issues/44347 was fixed in 2.10, 2.9, and 2.8 as a partial fix - #47126 is being worked on to wrap it up completely, which should encapsulate the problem outlined here. Closing this issue in favor of 47126
